src/Traits/VersionsCacheTrait.php

```diff
@@ -0,0 +1,197 @@
+<?php
+
+namespace SilverStripe\Versioned\Traits;
+
+use SilverStripe\Core\Config\Config;
+use SilverStripe\ORM\DataObject;
+use SilverStripe\ORM\DB;
+use SilverStripe\Model\List\ArrayList;
+use SilverStripe\Versioned\Versioned;
+use SilverStripe\Versioned\Versioned_Version;
+
+/**
+ * Functionality to cache Versioned::Versions(), which is used for the <DataObject>_Versions tables
+ */
+trait VersionsCacheTrait
+{
+    /**
+     * In-memory cache of Versions i.e. the <DataObject>_Versions table
+     * @internal
+     */
+    private static array $versions_cache = [];
+
+    /**
+     * Used whether to use the Versions cache
+     * @internal
+     */
+    private static bool $useVersionsCache = false;
+
+    /**
+     * Get whether to use the Versions cache
+     */
+    public static function getUseVersionsCache(): bool
+    {
+        return self::class::$useVersionsCache;
+    }
+
+    /**
+     * Set whether to use the Versions cache
+     */
+    public static function setUseVersionsCache(bool $use): void
+    {
+        self::class::$useVersionsCache = $use;
+    }
+
+    /**
+     * Prepopulate the in-memory Versions cache, which are used for <DataObject>_Versions tables
+     * The values will be retrieved using an efficient `WHERE "ID" IN (<ids>)` SQL query
+     * Note method this should not to be confused with prepopulateVersionNumberCache()
+     */
+    public static function prepopulateVersionsCache(
+        string $baseDataClass,
+        array $ids,
+        string $filter = '',
+        string $sort = '',
+        string $limit = '',
+        string $join = '',
+    ): void {
+        self::class::readVersions($baseDataClass, $ids, null, $filter, $sort, $limit, $join);
+    }
+
+    /**
+     * Shared internal logic which allows passing either a single ID for Versioned::Versions()
+     * or passing an array of IDs for VersionsCachingTrait::prepopulateVersionsCache()
+     *
+     * @param int|array $idOrIds A single record ID, or an array of record IDs
+     * @return ArrayList<Versioned_Version>|array<ArrayList<Versioned_Version>> an ArrayList when $idOrIds is an int, an array when $idOrIds is an array
```

Comment:
```suggestion
     * @return ArrayList<Versioned_Version>|array<int, ArrayList<Versioned_Version>> an ArrayList when $idOrIds is an int, an array when $idOrIds is an array
```
The IDs as a key in the array is important part of the return value

Reply:
Updated